library/packages/test/y2packager/product_upgrade_test.rb

```diff
@@ -96,4 +102,47 @@
       end
     end
   end
+
+  describe ".obsolete_upgrades" do
+    # upgrade from SLE12-SP3 + SUMA Proxy 3.2 + SUMA Branch Server 3.2
+    # to SLE15-SP1 (actually to SUMA Branch Server 4.0)
+    context "SUSE Manager 3.2 upgrade" do
+      let(:suma_products) { YAML.load_file(File.join(__dir__, "../data/zypp/products_update_suma.yml")) }
+
+      before do
+        allow(Yast::Pkg).to receive(:ResolvableProperties).with("", :product, "")
+          .and_return(suma_products)
+      end
+
+      it "returns obsoleted SLES + SUMA Proxy product" do
+        allow(Yast::Pkg).to receive(:ResolvableProperties).with("SUSE-Manager-Proxy", :product, "")
+          .and_return(suma_products.select { |p| p["name"] == "SUSE-Manager-Proxy" })
+        allow(Yast::Pkg).to receive(:ResolvableProperties).with("SLES", :product, "")
+          .and_return(suma_products.select { |p| p["name"] == "SLES" })
+        expect(described_class.obsolete_upgrades).to contain_exactly("SLES", "SUSE-Manager-Proxy")
+      end
+
+      it "returns empty list if the old product is removed by user" do
+        proxy_products = suma_products.select { |p| p["name"] == "SUSE-Manager-Proxy" }
+        # make it removed by user
+        proxy_products.find { |p| p["status"] == :removed }["transact_by"] = :user
+        allow(Yast::Pkg).to receive(:ResolvableProperties).with("SUSE-Manager-Proxy", :product, "")
+          .and_return(proxy_products)
+        allow(Yast::Pkg).to receive(:ResolvableProperties).with("SLES", :product, "")
+          .and_return(suma_products.select { |p| p["name"] == "SLES" })
+        expect(described_class.obsolete_upgrades).to eq([])
+      end
+    end
+  end
+
+  describe ".remove_obsolete_upgrades" do
+    it "marks the obsolete products for removal" do
+      obsolete = ["SLES", "SUSE-Manager-Proxy"]
+
+      expect(described_class).to receive(:obsolete_upgrades).and_return(obsolete)
+      obsolete.each { |o| expect(Yast::Pkg).to receive(:ResolvableRemove).with(o, :product) }
+
+      described_class.remove_obsolete_upgrades
+    end
+  end
```

Comment:
test for that bug fix of detection ( that reverse change )?

Reply:
OK, test added.